objects/SetupChecker/script.lua

```diff
@@ -1925,39 +1966,19 @@ function gainSpirit(player)
     end
 
     local count = 0
-    for i = 1,4 do
-        if not options[i] then
+    for i = 1, 4 do
+        if choices[i] == nil or choices[i].spirit == nil then
             local spirit, aspect = getNewSpirit(tags, complexities)
             if spirit then
                 count = count + 1
-                local label = spirit.getName()
-                if aspect ~= nil and aspect ~= "" then
-                    label = label.." - "..aspect
-                end
-                if i > numButtons then
-                    obj.createButton({
-                        click_function = "pickSpirit" .. i,
-                        function_owner = self,
-                        label = label,
-                        position = Vector(0,3,-5 + 2*i),
-                        rotation = Vector(0,0,0),
-                        scale = Vector(2/2.3,1,2/1.42),
-                        width = 4850,
-                        height = 600,
-                        font_size = 275,
-                    })
-                else
-                    obj.editButton({
-                        index = i - 1,
-                        label = label,
-                        width = 4850,
-                        height = 600,
-                    })
-                end
+                choices[i] = {spirit = spirit.getName(), aspect = aspect}
+            else
+                break
```

Comment:
this else case smells bad to me, it might not actually be causing issues right now, but also has potential with some code refactors to just break things

Reply:
The else is here to shortcut running `getNewSpirit` again on the same frame, after it has returned `nil` once.

I guess it *might* do that, given how it retries picking a spirit randomly that matches the criteria, but that seems unlikely enough to not worry about it? I think I'll look at changing that code to build a list of all pickable spirits (probably ahead of time when the checkboxes are changed), and use that, so that the `getNewSpirit` doesn't need to ever retry.

In any case, as in `replaceSpirit`, the `break` here is only an optimization, and shouldn't affect the correctness either way.

Reply:
generating a list of spirits ahead of time isn't the greatest idea imo since spirits can be added or removed from the mod on the fly, so you'd want to regenerate the list every time the spirit count changed

Reply:
> generating a list of spirits ahead of time isn't the greatest idea imo since spirits can be added or removed from the mod on the fly, so you'd want to regenerate the list every time the spirit count changed

It isn't all that hard to handle that: https://github.com/tomprince/spirit-island-mod/compare/custom-ui-for-gain-spirits...tomprince:spirit-island-mod:random-spirits